casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

```diff
@@ -148,20 +149,37 @@ object Proposer {
         // slashing
         preStateBonds <- RuntimeManager[F].computeBonds(preStateHash.toByteString)
         toSlash       = offenders intersect preStateBonds.filter { case (_, b) => b > 0 }.keySet
+        _             <- Log[F].info(s"Slashing senders: [${toSlash.map(_.show).mkString("; ")}]")
         // epoch
         changeEpoch = epochLength % nextBlockNum == 0
         // attestation
-        // no need to attest if nothing meaningful to finalize. TODO analyse finalization to suppress more
+        // no need to attest if nothing meaningful to finalize.
         dag <- BlockDagStorage[F].getRepresentation
         conflictSet = {
           val msgMap = dag.dagMessageState.msgMap
           parentHashes.flatMap(msgMap(_).seen) -- preState.fringe.flatMap(msgMap(_).seen)
         }
-        suppressAttestation <- conflictSet.toList
-                                .traverse(BlockStore[F].getUnsafe)
-                                .map(_.forall { b =>
-                                  b.state.systemDeploys.isEmpty && b.state.deploys.isEmpty
-                                })
+        nothingToFinalize = conflictSet.toList
+          .traverse(BlockStore[F].getUnsafe)
+          .map(_.forall { b =>
+            b.state.systemDeploys.isEmpty && b.state.deploys.isEmpty
+          })
+        waitingForSupermajorityToAttest = {
+          val newlySeen = creatorsLatestOpt
+            .map { prev =>
+              prev.justifications.flatMap(dag.dagMessageState.msgMap(_).seen) --
+                parentHashes.flatMap(dag.dagMessageState.msgMap(_).seen)
+            }
+            .getOrElse(Set())
+          newlySeen.toList.traverse(BlockStore[F].getUnsafe).map { newBlocks =>
+            val newStateTransition =
+              newBlocks.exists(b => b.state.systemDeploys.nonEmpty || b.state.deploys.nonEmpty)
+            val attestationStake =
+              preStateBonds.filterKeys(newBlocks.map(_.sender).toSet).values.toList.sum
+            !newStateTransition && attestationStake * 3 < preStateBonds.values.toList.sum * 2
+          }
+        }
+        suppressAttestation <- nothingToFinalize ||^ waitingForSupermajorityToAttest
         // push dummy deploy if needed
```

Comment:
What is the purpose of dummy deploys?

Reply:
There is a `dev` mode in the node when you can specify to put deploy in each block, signed with specific key. So you can create blocks without external orchestration, pretending that there is always deploy in the pool.

Reply:
From our standup discussion, we should later move this in a separate debugging code so it's not mixed with main block creation logic.